File: wasm/test/run_vectors.mjs
#!/usr/bin/env node
/* Golden-vector gate for the WASM target: render the synthetic vectors through
 * ae3synth.wasm + the JS binding and compare SHA-256 against tests/golden.sha256
 * -- the SAME golden file the native gate (tests/run_vectors.py) checks, so a
 * pass here proves WASM == native on the vectors, not merely WASM == itself.
 *
 * Expects dist/ae3synth.wasm built (make -C wasm) and regenerates the vector
 * assets via tests/make_vectors.py (deterministic), like the native gate. */
import { createHash } from "node:crypto";
import { spawnSync } from "node:child_process";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { renderWav } from "./render.mjs";

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = join(HERE, "../..");
const VEC = join(ROOT, "tests/vectors");

/* name -> render options; mirrors RENDERS in tests/run_vectors.py */
const RENDERS = {
    loop_seam: {},
    loop_seam_x2: { loop: 2 },
    drum_kit: {},
    bend: {},
    reverb_flag: {},
    adsr_edges: {},
    lfo: {},
    /* M8 cue layer -- mirrors run_vectors.py's --cue/--duck args */
    cue: { cueScale: 0.42, ducks: [{ which: 0, t0: 0.5, t1: 1.5 },
                                   { which: 1, t0: 1.0, t1: 2.2 }] },
};

/* vectors on a bank other than vec.hd (all share vec.bd); mirrors BANKS */
const BANKS = { lfo: "vlfo.hd" };

const gen = spawnSync("python3", [join(ROOT, "tests/make_vectors.py")]);
if (gen.status !== 0) {
    console.error(`make_vectors.py failed:\n${gen.stderr}`);
    process.exit(1);
}

const golden = {};
for (const line of readFileSync(join(ROOT, "tests/golden.sha256"), "utf8").trim().split("\n")) {
    const [h, name] = line.split(/\s+/);
    golden[name] = h;
}

const wasmBytes = readFileSync(join(HERE, "../dist/ae3synth.wasm"));
const hd = readFileSync(join(VEC, "vec.hd"));
const bd = readFileSync(join(VEC, "vec.bd"));

let fail = 0;
function judge(name, got) {
    const want = golden[name];
    const ok = got === want;
    console.log(`${ok ? "PASS" : "FAIL"} ${name}`
        + (ok ? "" : `  got ${got.slice(0, 16)}… want ${(want ?? "MISSING").slice(0, 16)}…`));
    fail += !ok;
}

function expect(cond, what) {
    if (!cond) {
        console.log(`FAIL ${what}`);
        fail++;
    }
}

for (const name of Object.keys(RENDERS).sort()) {
    const mid = readFileSync(join(VEC, (name.endsWith("_x2") ? name.slice(0, -3) : name) + ".mid"));
    const bankHd = BANKS[name] ? readFileSync(join(VEC, BANKS[name])) : hd;
    const { wav } = await renderWav({ wasmBytes, hd: bankHd, bd, mid, ...RENDERS[name] });
    judge(name, createHash("sha256").update(wav).digest("hex"));
}

/* EXST stream vectors: decode through the AE3Exst binding and frame as WAV --
 * the same golden entries the native exstdump renders hash to, so wasm decode
 * == native decode == the `ae3 exst --decode` framing. The parsed header
 * fields are hard-asserted against the authored synthetic values (this is the
 * ae3w_exst_parse flattener's gate). */
{
    const { AE3Exst } = await import("../js/ae3synth.mjs");
    const { wavHeader } = await import("../js/wav.mjs");
    const exst = await AE3Exst.instantiate(wasmBytes);
    const EXSTS = {
        exst_mono: { file: "vec_mono.x", trimPad: false },
        exst_mono_trim: { file: "vec_mono.x", trimPad: true },
        exst_stereo: { file: "vec_stereo.x", trimPad: false },
    };
    for (const [name, cfg] of Object.entries(EXSTS).sort()) {
        const data = readFileSync(join(VEC, cfg.file));
        const { header, pcm } = exst.decodeFile(data, { trimPad: cfg.trimPad });
        const wav = Buffer.concat([
            wavHeader(pcm.byteLength, header.channels, header.rate),
            new Uint8Array(pcm.buffer, pcm.byteOffset, pcm.byteLength),
        ]);
        judge(name, createHash("sha256").update(wav).digest("hex"));
    }
    const hm = exst.parseHeader(readFileSync(join(VEC, "vec_mono.x")));
    const hs = exst.parseHeader(readFileSync(join(VEC, "vec_stereo.x")));
    expect(hm.channels === 1 && hm.rate === 24000 && hm.length === 2
        && hm.loop === 0 && hm.loop_start === 0, "mono fields");
    expect(hm.vol_l[0] === 0x407F && hm.vol_r[0] === 0x407F, "mono volumes");
    expect(hs.channels === 2 && hs.rate === 48000 && hs.length === 4,
        "stereo fields (authored overstated length)");
    expect(hs.vol_l[0] === 0x407F && hs.vol_l[1] === 0 && hs.vol_r[0] === 0
        && hs.vol_r[1] === 0x407F && hs.reverb.every(x => x === 0),
        "stereo hard-pan + reverb clear");
}

/* decode_api: the bank-introspection surface through the binding, in the same
 * projection tests/run_vectors.py derives from wavdump --decode (its W lines
 * minus the seam-only n2=/hash2= fields). Same golden entry on both sides ->
 * WASM enumeration, table fields and PCM bytes == native. */
{
    const s = await (await import("../js/ae3synth.mjs")).AE3Synth.instantiate(wasmBytes);
    s.loadBank(hd, bd);
    const BASIS = 0xcbf29ce484222325n, PRIME = 0x100000001b3n, M64 = (1n << 64n) - 1n;
    let text = "";
    for (let i = 0; i < s.bankWaveforms(); i++) {
        const w = s.bankWaveform(i);
        const pcm = s.bankWaveformPcm(i);
        let h = BASIS;
        for (let k = 0; k < pcm.length; k++) {
            const v = pcm[k] & 0xffff;
            h = ((h ^ BigInt(v & 0xff)) * PRIME) & M64;
            h = ((h ^ BigInt(v >> 8)) * PRIME) & M64;
        }
        text += `W addr=${w.addr} n=${w.samples} loop=${w.loop_start}`
              + ` hash=${h.toString(16).padStart(16, "0")} prog=${w.prog}`
              + ` tone=${w.tone} root=${w.root} tune=${w.tune} refs=${w.refs}\n`;
    }
    s.dispose();
    judge("decode_api", createHash("sha256").update(text).digest("hex"));
}

/* source_state: the live per-voice source coordinate through the public JS
 * binding. The text is byte-identical to serender --source-state and shares
 * its native golden entry. */
{
    const {
        AE3Synth, VOICE_STATE_SIZE,
    } = await import("../js/ae3synth.mjs");
    const s = await AE3Synth.instantiate(wasmBytes);
    s.loadBank(readFileSync(join(VEC, "vec_se.hd")), bd);
    const envNames = ["ATTACK", "DECAY", "SUSTAIN", "RELEASE", "OFF"];
    const sourceNames = ["NONE", "ONESHOT", "LOOPED", "NOISE"];
    const mix = new Float32Array(2);
    let text = "", frame = 0;
    const line = (tag, slot, state, at = frame) => {
        text += `${tag} slot=${slot} frame=${at} in_use=${Number(state.in_use)}`
              + ` active=${Number(state.active)} released=${Number(state.released)}`
              + ` key=${state.key} env=${state.env} env_phase=${envNames[state.env_phase]}`
              + ` se_prog=${state.se_prog} source=${sourceNames[state.source_kind]}`
              + ` waveform=${state.waveform} samples=${state.source_samples}`
              + ` loop_start=${state.source_loop_start}`
              + ` phase_q12=${state.source_phase_q12} loops=${state.source_loops}\n`;
    };
    const activeSlot = key => {
        for (let slot = 0; slot < 48; slot++) {
            const state = s.voice(slot);
            if (state.active && state.key === key) return slot;
        }
        return -1;
    };

    line("idle", 0, s.voice(0));
    text += `out_of_range=${Number(s.voice(48) !== null)}\n`;
    expect(VOICE_STATE_SIZE === 12, "source_state voice stride");

    s.noteOn(0, 60, 100);
    const loopSlot = activeSlot(60);
    let current = s.voice(loopSlot);
    line("loop_unprimed", loopSlot, current);
    const packed = new Float64Array(VOICE_STATE_SIZE);
    expect(s.voiceInto(loopSlot, packed)
        && packed[0] === 3 && packed[8] === 224 && packed[9] === 56
        && packed[10] === 0, "source_state packed layout");
    s.render(mix, 1);
    frame++;
    current = s.voice(loopSlot);
    line("loop_primed", loopSlot, current);

    let previous = current;
    while (current.source_loops < 2 && frame < 2000) {
        s.render(mix, 1);
        frame++;
        current = s.voice(loopSlot);
        if (current.source_loops !== previous.source_loops) {
            line(current.source_loops === 1 ? "loop1_before" : "loop2_before",
                 loopSlot, previous, frame - 1);
            line(current.source_loops === 1 ? "loop1_after" : "loop2_after",
                 loopSlot, current);
        }
        previous = current;
    }
    expect(current.source_loops === 2, "source_state second seam");

    s.noteOff(0, 60);
    current = s.voice(loopSlot);
    line("loop_release", loopSlot, current);

    s.noteOn(0, 61, 100);
    const noiseSlot = activeSlot(61);
    line("noise", noiseSlot, s.voice(noiseSlot));

    s.noteOn(0, 62, 100);
    const oneSlot = activeSlot(62);
    current = s.voice(oneSlot);
    line("oneshot", oneSlot, current);
    while (current.active && frame < 5000) {
        s.render(mix, 1);
        frame++;
        current = s.voice(oneSlot);
    }
    expect(!current.active, "source_state one-shot end");
    line("ended", oneSlot, current);
    s.dispose();

    const parsed = Object.fromEntries(text.trim().split("\n")
        .filter(row => !row.startsWith("out_of_range="))
        .map(row => {
            const fields = row.split(" ");
            return [fields[0], Object.fromEntries(fields.slice(1)
                .map(field => field.split("=", 2)))];
        }));
    expect(parsed.loop_unprimed.samples === "224"
        && parsed.loop_unprimed.loop_start === "56", "source_state source metadata");
    expect(Number(parsed.loop1_before.phase_q12) > Number(parsed.loop1_after.phase_q12)
        && parsed.loop1_after.loops === "1" && parsed.loop2_after.loops === "2",
        "source_state phase wraps");
    expect(parsed.loop1_before.env === parsed.loop1_after.env
        && parsed.loop1_before.env_phase === parsed.loop1_after.env_phase,
        "source_state envelope continuity");
    expect(parsed.noise.source === "NOISE" && parsed.noise.phase_q12 === "-1",
        "source_state noise sentinel");
    expect(parsed.oneshot.source === "ONESHOT" && parsed.oneshot.loop_start === "-1",
        "source_state one-shot sentinel");
    expect(parsed.idle.source === "NONE" && parsed.ended.env_phase === "OFF",
        "source_state inactive sentinels");
    expect(text.includes("out_of_range=0\n"), "source_state out-of-range query");
    judge("source_state", createHash("sha256").update(text).digest("hex"));
}

process.exit(fail ? 1 : 0);

File: tests/run_vectors.py
#!/usr/bin/env python3
"""Golden-vector gate: render the synthetic vectors, compare SHA-256 to golden.

Usage: python3 tests/run_vectors.py [--update]
Builds the harness (make), regenerates the vector assets from make_vectors.py
(they are deterministic), renders each through wavdump with no IRX donors
(ET pitch table, dry mix), and byte-hash-compares against golden.sha256.
--update rewrites golden.sha256 from the current renders instead.
"""
import hashlib
import os
import subprocess
import sys
import tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
VEC = os.path.join(HERE, "vectors")
WAVDUMP = os.path.join(ROOT, "harness", "wavdump")
EXSTDUMP = os.path.join(ROOT, "harness", "exstdump")
SERENDER = os.path.join(ROOT, "harness", "serender")
GOLDEN = os.path.join(HERE, "golden.sha256")

# name -> extra wavdump args
RENDERS = {
    "loop_seam": [],
    "loop_seam_x2": ["--loop", "2"],
    "drum_kit": [],
    "bend": [],
    "reverb_flag": [],
    "adsr_edges": [],
    "lfo": [],
    # M8 cue layer: authored scale + both duck groups, overlapping windows
    # (x0.49), ramps in both directions, edges on tick boundaries
    "cue": ["--cue", "0.42", "--duck", "0:0.5:1.5", "--duck", "1:1.0:2.2"],
}

# vectors rendered against a bank other than vec.hd (all share vec.bd)
BANKS = {"lfo": "vlfo.hd"}

# Embedded-SE sequencer: exact 480 Hz and console 60 Hz dispatch, plus host
# play-once/forever control of an authored count=0 jump.
# name -> (bank vector, extra serender args)
SE_RENDERS = {
    "se_exact": ("vec_se.hd", []),
    "se_tick": ("vec_se.hd", ["--tick-events"]),
    "se_loop_once": ("vec_se_inf.hd", ["--loop", "0"]),
    "se_loop_forever": ("vec_se_inf.hd", ["--loop", "127"]),
}

# EXST stream vectors through exstdump --decode, hashed as whole WAVs. The
# wasm gate rebuilds the identical files through the AE3Exst binding, so the
# shared golden entries prove WASM decode == native decode == the framing
# `ae3 exst --decode` writes. name -> (vector, extra args)
EXSTS = {
    "exst_mono": ("vec_mono.x", []),
    "exst_mono_trim": ("vec_mono.x", ["--trim-pad"]),
    "exst_stereo": ("vec_stereo.x", []),   # header overstates length: warns
}

# stdout modes, hashed as text. "decode" gates the bank-introspection API:
# wavdump --decode enumerates waveforms through it, cross-checks the table and
# PCM against the raw decoder stream internally, and prints per-waveform
# sample counts / loop points / FNV hashes / naming refs.
DUMPS = {
    "decode": ["--decode"],
}


def source_states(text):
    states = {}
    for line in text.decode().splitlines():
        fields = line.split()
        if not fields or "=" in fields[0]:
            continue
        states[fields[0]] = {
            key: value for key, value in (field.split("=", 1) for field in fields[1:])
        }
    return states


def validate_source_state(text):
    state = source_states(text)

    def number(tag, field):
        return int(state[tag][field])

    def need(condition, message):
        if not condition:
            raise RuntimeError(f"source-state assertion failed: {message}")

    need(number("loop_unprimed", "samples") == 224, "source length")
    need(number("loop_unprimed", "loop_start") == 56, "loop start")
    need(number("loop_unprimed", "phase_q12") == 0, "unprimed phase")
    need(number("loop_primed", "phase_q12") > 0, "primed phase")
    need(number("loop1_before", "phase_q12") > number("loop1_after", "phase_q12"),
         "first phase wrap")
    need(number("loop1_before", "loops") == 0
         and number("loop1_after", "loops") == 1
         and number("loop2_after", "loops") == 2, "seam counts")
    need(number("loop1_before", "env") == number("loop1_after", "env")
         and state["loop1_before"]["env_phase"] == state["loop1_after"]["env_phase"],
         "envelope seam continuity")
    need(state["loop_release"]["env_phase"] == "RELEASE"
         and number("loop_release", "loops") == 2, "release continuity")
    need(state["noise"]["source"] == "NOISE"
         and number("noise", "waveform") == -1
         and number("noise", "phase_q12") == -1, "noise sentinel")
    need(state["oneshot"]["source"] == "ONESHOT"
         and number("oneshot", "loop_start") == -1
         and number("oneshot", "phase_q12") == 0, "one-shot sentinel")
    for tag in ("idle", "ended"):
        need(state[tag]["source"] == "NONE"
             and state[tag]["env_phase"] == "OFF"
             and number(tag, "waveform") == -1
             and number(tag, "phase_q12") == -1, f"{tag} sentinel")
    need(b"out_of_range=0\n" in text, "out-of-range query")


def main():
    update = "--update" in sys.argv
    subprocess.run(["make", "-C", os.path.join(ROOT, "core")],
                   check=True, capture_output=True)
    subprocess.run(["make", "-C", os.path.join(ROOT, "harness"),
                    "wavdump", "exstdump", "serender"],
                   check=True, capture_output=True)
    subprocess.run([sys.executable, os.path.join(HERE, "make_vectors.py")],
                   check=True, capture_output=True)

    hashes = {}
    with tempfile.TemporaryDirectory() as td:
        for name, extra in RENDERS.items():
            mid = name[:-3] if name.endswith("_x2") else name
            out = os.path.join(td, name + ".wav")
            subprocess.run(
                [WAVDUMP, os.path.join(VEC, BANKS.get(name, "vec.hd")),
                 os.path.join(VEC, "vec.bd"),
                 os.path.join(VEC, mid + ".mid"), *extra, "-o", out],
                check=True, capture_output=True)
            with open(out, "rb") as f:
                hashes[name] = hashlib.sha256(f.read()).hexdigest()

        for name, (bank, extra) in SE_RENDERS.items():
            out = os.path.join(td, name + ".wav")
            subprocess.run(
                [SERENDER, "--seconds", "3", *extra, "-o", out,
                 os.path.join(VEC, bank), os.path.join(VEC, "vec.bd"),
                 "0", "0"],
                check=True, capture_output=True)
            with open(out, "rb") as f:
                hashes[name] = hashlib.sha256(f.read()).hexdigest()

        for name, (vec, extra) in EXSTS.items():
            out = os.path.join(td, name + ".wav")
            subprocess.run(
                [EXSTDUMP, "--decode", *extra, os.path.join(VEC, vec),
                 "-o", out],
                check=True, capture_output=True)
            with open(out, "rb") as f:
                hashes[name] = hashlib.sha256(f.read()).hexdigest()

    for name, flags in DUMPS.items():
        r = subprocess.run(
            [WAVDUMP, *flags, os.path.join(VEC, "vec.hd"), os.path.join(VEC, "vec.bd")],
            check=True, capture_output=True)
        hashes[name] = hashlib.sha256(r.stdout).hexdigest()
        if name == "decode":
            # API-visible projection (no n2=/hash2= seam fields): the wasm gate
            # reconstructs this exact text through the JS binding, so the shared
            # golden entry proves WASM API == native API on the vectors.
            api = "".join(
                " ".join(t for t in line.split()
                         if not t.startswith(("n2=", "hash2="))) + "\n"
                for line in r.stdout.decode().splitlines())
            hashes["decode_api"] = hashlib.sha256(api.encode()).hexdigest()

    source = subprocess.run(
        [SERENDER, "--source-state", os.path.join(VEC, "vec_se.hd"),
         os.path.join(VEC, "vec.bd")],
        check=True, capture_output=True).stdout
    validate_source_state(source)
    hashes["source_state"] = hashlib.sha256(source).hexdigest()
    if update:
        with open(GOLDEN, "w") as f:
            for name, h in sorted(hashes.items()):
                f.write(f"{h}  {name}\n")
        print(f"golden.sha256 updated ({len(hashes)} renders)")
        return 0

    golden = {}
    with open(GOLDEN) as f:
        for line in f:
            h, name = line.split()
            golden[name] = h
    fail = 0
    for name in sorted(hashes):
        got, want = hashes[name], golden.get(name)
        ok = got == want
        print(f"{'PASS' if ok else 'FAIL'} {name}"
              + ("" if ok else f"  got {got[:16]}… want {(want or 'MISSING')[:16]}…"))
        fail += not ok
    extra = set(golden) - set(hashes)
    if extra:
        print(f"FAIL stale golden entries: {sorted(extra)}")
        fail += 1
    return 1 if fail else 0


if __name__ == "__main__":
    sys.exit(main())
